### PR TITLE
Add and integrate `krel anago push`

### PR DESCRIPTION
**Pull request by saschagrunert**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:
Target of this subcommand is to completely replace the push logic in
anago. For now we only stage the local artifacts with it to verify that
everything works as intended.
#### Special notes for your reviewer:
Requires #1596 
/hold
#### Does this PR introduce a user-facing change?

```release-note
Added `krel anago push` subcommand for publishing anago stage and release artifacts.
```

### CODE REVIEW

**Comment by justaugustus** [comment]
@saschagrunert -- Needs a rebase now that the directory changes in https://github.com/kubernetes/release/pull/1596 have merged.

**Comment by saschagrunert** [comment]
> @saschagrunert -- Needs a rebase now that the directory changes in #1596 have merged.

Yes, rebased but this still needs to hold having to succeed a test run.

**Comment by saschagrunert** [comment]
Comparing the already existing staged output:
- https://console.cloud.google.com/storage/browser/kubernetes-release-gcb/stage/v1.19.2-rc.0.12%2B19706d90d87784/v1.19.2/gcs-stage/v1.19.2?pageState=(%22StorageObjectListTable%22:(%22f%22:%22%255B%255D%22))&prefix=&forceOnObjectsSortingFiltering=false

With the run:
```
> export TOOL_ORG=saschagrunert && \
  export TOOL_BRANCH=krel-anago-push && \
  go run ./cmd/krel gcbmgr --stage --branch master --type alpha
```
Logs: https://console.cloud.google.com/cloud-build/builds/8aae5691-760e-49c3-9133-b7413693391c?project=kubernetes-release-test

Staged output:
- :hourglass: wait for result

**Comment by saschagrunert** [comment]
Ready for review.

/hold cancel

File: cmd/krel/cmd/anago/push.go [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+
+// pushCmd represents the subcommand for `krel anago push`
+var pushCmd = &cobra.Command{
+	Use:   "push",
+	Short: "Push release artifacts into the Google Cloud",
+	Long: `krel anago push
+
+This subcommand can be used to push the release artifacts to the Google Cloud. 
+It's only indented to be used from anago, which means the command might be
+removed in future releases again if anago goes end of life.
```
Comment (cpanato) [inline cmd/krel/cmd/anago/push.go:34]: maybe `... when anago goes end of life` 
non blocking

**Review by cpanato — APPROVED** [review]
small non blocking comment, thanks for this!!
/lgtm
/hold if others want to review

File: pkg/util/common.go [hunk 2]
```diff
@@ -442,81 +442,97 @@ func SemverToTagString(tag semver.Version) string {
...
 
 // CopyDirContentsLocal copies local directory contents from one local location
 // to another.
 func CopyDirContentsLocal(src, dst string) error {
+	logrus.Infof("Trying to copy dir %s to %s", src, dst)
 	// If initial destination does not exist create it.
 	if _, err := os.Stat(dst); err != nil {
-		if err := os.MkdirAll(dst, os.FileMode(0755)); err != nil {
-			return errors.Wrapf(err, "Unable to create directory at path %s", dst)
+		if err := os.MkdirAll(dst, os.FileMode(0o755)); err != nil {
```
Comment (MushuEE) [inline pkg/util/common.go:490]: I haven't seen the 0o### syntax and not seeing it in a quick search. Is that a new standard?

File: pkg/release/push.go [hunk 3]
```diff
@@ -288,60 +279,91 @@ func (p *PushBuild) findLatestVersion() (latestVersion string, err error) {
 	if !valid {
 		return "", errors.Errorf("build version %s is not valid for release", latestVersion)
 	}
+
+	if p.opts.CI && IsDirtyBuild(latestVersion) {
+		return "", errors.New("refusing to push dirty build with --ci flag given")
```
[the remark this answers is not in the document]
Comment (saschagrunert) [reply pkg/release/push.go]: Fixed as suggested :+1:

File: pkg/release/release.go [hunk 4]
```diff
@@ -322,54 +318,60 @@ func GetOCIManifest(tarPath string) (*ocispec.Manifest, error) {
...
-				"Skipping platform %q because it's not a directory",
-				platform.Name(),
+				"Skipping platform and arch %q because it's not a directory",
+				platformArch.Name(),
 			)
 			continue
 		}
 
-		logrus.Infof("Copying binaries for %s platform", platform.Name())
+		split := strings.SplitN(platformArch.Name(), "-", 2)
```
[the remark this answers is not in the document]
Comment (saschagrunert) [reply pkg/release/release.go]: Sure, changed as suggested :+1:

File: pkg/util/common.go (hunk 2 above)
Comment (saschagrunert) [reply pkg/util/common.go:490]: Yes, go 1.13 added the octal literals: https://golang.org/ref/spec#Integer_literals

File: pkg/release/push.go [hunk 5]
```diff
@@ -288,60 +279,91 @@ func (p *PushBuild) findLatestVersion() (latestVersion string, err error) {
...
+// copyStageFiles takes the staging dir and copies each file of `files` into
+// it. It also ensures that the base dir exists before copying the file (if the
+// file is `required`).
+func (p *PushBuild) copyStageFiles(stageDir string, files []stageFile) error {
+	for _, file := range files {
+		dstPath := filepath.Join(stageDir, file.dstPath)
+
+		if file.required {
+			if err := os.MkdirAll(
+				filepath.Dir(dstPath), os.FileMode(0o755),
```
[the remark this answers is not in the document]
Comment (saschagrunert) [reply pkg/release/push.go:359]: Hm, I don't think that we skipped the files previously. It was just not working at all :)

File: cmd/krel/cmd/anago/push.go (hunk 1 above)
Comment (saschagrunert) [reply cmd/krel/cmd/anago/push.go]: Fixed as suggest :+1:

**Comment by saschagrunert** [comment]
PTAL again @kubernetes/release-engineering

**Comment by puerco** [comment]
It looks good to me, I've been following the discussion and I'm in favor of lifting the hold. 

/lgtm

**Comment by saschagrunert** [comment]
@justaugustus @tpepper please take a look for the blocked paths.

**Comment by justaugustus** [comment]
Here we go!
/lgtm
/approve

**Pull request merged**